### PR TITLE
Multiple x:Bind Updates (TwoWay, x:DefaultBindMode, path less bindings)

### LINKED ISSUE #2587: TwoWay x:Bind doesn't work
## Current behavior

TwoWay x:Bind doesn't work even without BindBack property.

## Expected behavior

Synchronize value between source to target the same as UWP.

## How to reproduce it (as minimally and precisely as possible)

1. Create a ViewModel class that is implemented INotifyPropertyChanged interface.
2. Define a property type is the ViewModel.
3. Define two way x:Bind on the XAML such as `{x:Bind ViewModel.PropertyA, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}`.

## Environment

Nuget Package:
- Uno.UI

Package Version(s):
- Uno.UI 2.1.0-dev998

Affected platform(s):

- [x] iOS
- [x] Android
- [x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

Visual Studio:

- [ ] 2017 (version: )
- [x] 2019 (version: 16.4.3)
- [ ] for Mac (version: )

Relevant plugins:

- [ ] Resharper (version: )

## Anything else we need to know?

I created a repository for repro.
https://github.com/runceel/TowWayXBind

And I recorded behaviors on each platform:

### UWP
It works.
xbinduwp

### Android
It doesn't work.
xbinddroid

### WASM
It doesn't work.
xbindwasm

### LINKED ISSUE #2693: x:DefaultBindMode="OneWay" is not supported on all platforms
## Current behavior

x:DefaultBindMode="OneWay" has no effect for iOS/Android while it works perfectly for UWP [and WASM (as a bug at the moment)].
ios DefaultBindMode issue

## Expected behavior

x:DefaultBindMode="OneWay"  should work
UWP normal behavior for DefaultBindMode

## How to reproduce it (as minimally and precisely as possible)
Repro:
https://github.com/zipswich/AppDebugUno

## Environment

Nuget Package:
Uno.UI

Package Version(s):
2.1.0-dev.1338

Affected platform(s):

- [x] iOS
- [x] Android
- [x] WebAssembly (working as a bug at the moment)
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

Visual Studio:

- [ ] 2017 (version: )
- [x] 2019 (version: )
- [ ] for Mac (version: )

Relevant plugins:

- [ ] Resharper (version: )

### LINKED ISSUE #2694: x:Bind Converter needs a dummy value to be called on iOS
## Current behavior

sDummy is needed for the following TextBlock to call TextConverterDebug
```
            <TextBlock
                Margin="10 0 0 0"
                FontSize="30"
                Text="{x:Bind sDummy, ConverterParameter=test, Converter={StaticResource TextConverterDebug}}"
                 />
```
On iOS:
image

## Expected behavior

Converters should be called regardless of whether the value is present like the behavior of the UWP head:
image

## How to reproduce it (as minimally and precisely as possible)

https://github.com/zipswich/AppDebugUno

## Environment

Nuget Package:
Uno.UI
Package Version(s):
2.1.0-dev.1338
Affected platform(s):

- [x] iOS
- [ ] Android
- [ ] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

Visual Studio:

- [ ] 2017 (version: )
- [x] 2019 (version: )
- [ ] for Mac (version: )

Relevant plugins:

- [ ] Resharper (version: )

### PR DESCRIPTION
**Pull request by jeromelaban**
GitHub Issue (If applicable): Fixes #2587, Fixes #2693, Fixes #2694

## PR Type

What kind of change does this PR introduce?
- Bugfix
- Feature

## What is the new behavior?

- Fix support for `x:Bind` with `Mode=TwoWay`
- Add support for `x:DefaultBindMode`
- Fix `x:Bind` without a specifide `Path`

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](https://github.com/unoplatform/uno/blob/master/README.md#uno-features)
- [x] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] [Wasm UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Updated the [Release Notes](https://github.com/unoplatform/uno/tree/master/doc/ReleaseNotes)
- [x] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Data/BindingHelper.cs [hunk 1]
```diff
@@ -8,9 +8,11 @@ namespace Uno.UI.Xaml
 	public static class BindingHelper
 	{
 		public static Binding SetBindingXBindProvider(Binding binding, object compiledSource, Func<object, object> xBindSelector, string[] propertyPaths = null)
-		{
-			binding.SetBindingXBindProvider(compiledSource, xBindSelector, propertyPaths);
+			=> SetBindingXBindProvider(binding, compiledSource, xBindSelector, null, propertyPaths);
 
+		public static Binding SetBindingXBindProvider(Binding binding, object compiledSource, Func<object, object> xBindSelector, Action<object, object> xBindBack, string[] propertyPaths = null)
+		{
+			binding.SetBindingXBindProvider(compiledSource, xBindSelector, xBindBack, propertyPaths);
```
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/Data/BindingHelper.cs]: Codacy Issue found: [Refactor this method to add validation of parameter 'binding' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs [hunk 2]
```diff
@@ -43,5 +43,103 @@ public void When_StateTrigger()
...
+			var myObject = SUT.FindName("myObject") as Binding_DataTemplate_TwoWayTestObject;
+
+			Assert.AreEqual(0, myObject.MyProperty);
+			Assert.AreEqual(0, rootData.MyIntProperty);
+
+			rootData.MyIntProperty = 1;
+
+			Assert.AreEqual(1, myObject.MyProperty);
+
+			Console.WriteLine("myObject.MyProperty = 2");
```
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs]: Codacy Issue found: [Remove this logging statement.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs [hunk 3]
```diff
@@ -43,5 +43,103 @@ public void When_StateTrigger()
...
+			SUT.ForceLoaded();
+
+			Assert.AreEqual(0, SUT.myObject.MyProperty);
+			Assert.AreEqual(0, SUT.MyIntProperty);
+
+			SUT.MyIntProperty = 1;
+
+			Assert.AreEqual(1, SUT.myObject.MyProperty);
+
+			Console.WriteLine("SUT.myObject.MyProperty = 2");
```
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs]: Codacy Issue found: [Remove this logging statement.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs [hunk 4]
```diff
@@ -43,5 +43,103 @@ public void When_StateTrigger()
...
+			var myObjectNestedProperty = SUT.FindName("myObjectNestedProperty") as Binding_DataTemplate_TwoWayTestObject;
+
+			Assert.AreEqual(0, myObjectNestedProperty.MyProperty);
+			Assert.AreEqual(0, rootData.Model.MyIntProperty);
+
+			rootData.Model.MyIntProperty = 1;
+
+			Assert.AreEqual(1, myObjectNestedProperty.MyProperty);
+
+			Console.WriteLine("myObjectNestedProperty.MyProperty = 2");
```
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs]: Codacy Issue found: [Remove this logging statement.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs [hunk 5]
```diff
@@ -43,5 +43,297 @@ public void When_StateTrigger()
...
+			SUT.ForceLoaded();
+
+			Assert.AreEqual("0", SUT.myObject.MyProperty);
+			Assert.AreEqual(0, SUT.MyIntProperty);
+
+			SUT.MyIntProperty = 1;
+
+			Assert.AreEqual("1", SUT.myObject.MyProperty);
+
+			Console.WriteLine("SUT.myObject.MyProperty = 2");
```
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs]: Codacy Issue found: [Remove this logging statement.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs [hunk 6]
```diff
@@ -43,5 +43,297 @@ public void When_StateTrigger()
...
+			var myObjectNestedProperty = SUT.FindName("myObjectNestedProperty") as Binding_TwoWay_Object_DataTemplate_TestObject;
+
+			Assert.AreEqual(0, myObjectNestedProperty.MyProperty);
+			Assert.AreEqual(0, rootData.Model.MyIntProperty);
+
+			rootData.Model.MyIntProperty = 1;
+
+			Assert.AreEqual(1, myObjectNestedProperty.MyProperty);
+
+			Console.WriteLine("myObjectNestedProperty.MyProperty = 2");
```
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs]: Codacy Issue found: [Remove this logging statement.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI/UI/Xaml/Data/BindingHelper.cs (hunk 1 above)
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/Data/BindingHelper.cs]: Codacy Issue found: [Refactor this method to add validation of parameter 'binding' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs [hunk 7]
```diff
@@ -43,5 +43,297 @@ public void When_StateTrigger()
...
+			var myObject = SUT.FindName("myObject") as Binding_TwoWay_Object_DataTemplate_TestObject;
+
+			Assert.AreEqual(0, myObject.MyProperty);
+			Assert.AreEqual(0, rootData.MyIntProperty);
+
+			rootData.MyIntProperty = 1;
+
+			Assert.AreEqual(1, myObject.MyProperty);
+
+			Console.WriteLine("myObject.MyProperty = 2");
```
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs]: Codacy Issue found: [Remove this logging statement.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack_DataTemplate.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		{
+			get => _myIntProperty;
+			set
+			{
+				_myIntProperty = value;
+				PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(MyIntProperty)));
+			}
+		}
+
+		public void MyBindBack(string value) => MyIntProperty = int.Parse(value);
```
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack_DataTemplate.xaml.cs]: Codacy Issue found: [Use the overload that takes a 'CultureInfo' or 'IFormatProvider' parameter.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs [hunk 9]
```diff
@@ -43,5 +43,297 @@ public void When_StateTrigger()
...
+			SUT.ForceLoaded();
+
+			Assert.AreEqual("0", SUT.myObjectNestedProperty.MyProperty);
+			Assert.AreEqual(0, SUT.Model.MyIntProperty);
+
+			SUT.Model.MyIntProperty = 1;
+
+			Assert.AreEqual("1", SUT.myObjectNestedProperty.MyProperty);
+
+			Console.WriteLine("SUT.myObjectNestedProperty.MyProperty = 2");
```
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Given_xBind_Binding.cs]: Codacy Issue found: [Remove this logging statement.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+		{
+			get => _myIntProperty;
+			set
+			{
+				_myIntProperty = value;
+				PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(MyIntProperty)));
+			}
+		}
+
+		public void MyBindBack(string value) => MyIntProperty = int.Parse(value);
```
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack.xaml.cs]: Codacy Issue found: [Use the overload that takes a 'CultureInfo' or 'IFormatProvider' parameter.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/SourceGenerators/Uno.UI.SourceGenerators.Tests/Given_XBindParser.cs [hunk 11]
```diff
@@ -37,9 +38,11 @@ bool IsStaticMethod(string name)
 
 				return false;
 			}
+
 			var props = XBindExpressionParser.ParseProperties(inputExpression, IsStaticMethod);
 
-			Assert.IsTrue(output.SequenceEqual(props), $"Expected [{string.Join(";", output)}], got [{string.Join(";", props)}]");
+			Assert.AreEqual(props.hasFunction, hasFunction, $"Expected hasFunction=true for [{inputExpression}]");
+			Assert.IsTrue(output.SequenceEqual(props.properties), $"Expected [{string.Join(";", output)}], got [{string.Join(";", props)}]");
```
Comment (carldebilly) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Tests/Given_XBindParser.cs]: Codacy Issue found: [Refactor this method to add validation of parameter 'output' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

Comment (carldebilly) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Tests/Given_XBindParser.cs]: Codacy Issue found: [Refactor this method to add validation of parameter 'output' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack_DataTemplate.xaml.cs (hunk 8 above)
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack_DataTemplate.xaml.cs]: Codacy Issue found: [Use the overload that takes a 'CultureInfo' or 'IFormatProvider' parameter.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack.xaml.cs (hunk 10 above)
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack.xaml.cs]: Codacy Issue found: [Use the overload that takes a 'CultureInfo' or 'IFormatProvider' parameter.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI/UI/Xaml/Data/BindingHelper.cs (hunk 1 above)
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/Data/BindingHelper.cs]: Codacy Issue found: [Refactor this method to add validation of parameter 'binding' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack_DataTemplate.xaml.cs (hunk 8 above)
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack_DataTemplate.xaml.cs]: Codacy Issue found: [Use the overload that takes a 'CultureInfo' or 'IFormatProvider' parameter.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack.xaml.cs (hunk 10 above)
Comment (carldebilly) [inline src/Uno.UI.Tests/Windows_UI_Xaml_Data/xBindTests/Controls/Binding_TwoWay_BindBack.xaml.cs]: Codacy Issue found: [Use the overload that takes a 'CultureInfo' or 'IFormatProvider' parameter.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/SourceGenerators/Uno.UI.SourceGenerators.Tests/Given_XBindParser.cs (hunk 11 above)
Comment (carldebilly) [inline src/SourceGenerators/Uno.UI.SourceGenerators.Tests/Given_XBindParser.cs]: Codacy Issue found: [Refactor this method to add validation of parameter 'output' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/Uno.UI/UI/Xaml/Data/BindingHelper.cs (hunk 1 above)
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/Data/BindingHelper.cs]: Codacy Issue found: [Refactor this method to add validation of parameter 'binding' before using it.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5087470)

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs [hunk 12]
```diff
@@ -18,405 +18,412 @@
 using Uno.Equality;
 using Uno.Logging;
 using Uno.UI.SourceGenerators.XamlGenerator.XamlRedirection;
+using Mono.Cecil;
```
Comment (davidjohnoliver) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs]: Is Cecil actually being used? (Or the using below)

Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs]: I don't know, it was from your commit I think :). Will check.

**Pull request merged**